src/core/lib/surface/call.c

```diff
@@ -1290,11 +1290,10 @@ static void receiving_stream_ready(grpc_exec_ctx *exec_ctx, void *bctlp,
     cancel_with_error(exec_ctx, call, STATUS_FROM_SURFACE,
                       GRPC_ERROR_REF(error));
   }
-  if (call->has_initial_md_been_received || error != GRPC_ERROR_NONE ||
-      call->receiving_stream == NULL) {
+  if (error != GRPC_ERROR_NONE || call->receiving_stream == NULL ||
+      !gpr_atm_rel_cas(&call->saved_receiving_stream_ready_bctlp, 0,
+                       (gpr_atm)bctlp)) {
     process_data_after_md(exec_ctx, bctlp);
```

Comment:
It looks like there may an unrelated bug here, which is that if the recv_initial_metadata and recv_message ops were in different batches, we will use the wrong batch_control structure in this call.  I think that just as we save the recv_message batch_control struct to be used in recv_initial_metadata_ready, we also need to save the recv_initial_metadata batch_control struct to be used in recv_message_ready.